File: Assets/Bundles/Shared/Resources/TS/Item/HeldItems/HeldItem.lua
-- Compiled with unity-ts v2.1.0-75
local SoundUtil = require("Shared/TS/Util/SoundUtil").SoundUtil
local _ReferenceManagerResources = require("Shared/TS/Util/ReferenceManagerResources")
local BundleGroupNames = _ReferenceManagerResources.BundleGroupNames
local ReferenceManagerAssets = _ReferenceManagerResources.ReferenceManagerAssets
local RunUtil = require("Shared/TS/Util/RunUtil").RunUtil
local TimeUtil = require("Shared/TS/Util/TimeUtil").TimeUtil
local HeldItem
do
	HeldItem = setmetatable({}, {
		__tostring = function()
			return "HeldItem"
		end,
	})
	HeldItem.__index = HeldItem
	function HeldItem.new(...)
		local self = setmetatable({}, HeldItem)
		return self:constructor(...) or self
	end
	function HeldItem:constructor(entity, newMeta)
		self.serverOffsetMargin = 0.025
		self.lastUsedTime = 0
		self.chargeStartTime = 0
		self.isCharging = false
		self.entity = entity
		self.meta = newMeta
		-- Load the asset references
		local _result = self.meta.itemAssets
		if _result ~= nil then
			_result = _result.assetBundleId
		end
		if _result ~= 0 and (_result == _result and _result) then
			local _bundleGroups = ReferenceManagerAssets.bundleGroups
			local _assetBundleId = self.meta.itemAssets.assetBundleId
			self.bundles = _bundleGroups[_assetBundleId]
		end
	end
	function HeldItem:Log(message, isError)
		if isError == nil then
			isError = false
		end
		return nil
	end
	function HeldItem:OnEquip()
		self:Log("OnEquip")
		-- Load that items animations and play equip animation
		local _result = self.entity.anim
		if _result ~= nil then
			local _result_1 = self.meta.itemAssets
			if _result_1 ~= nil then
				_result_1 = _result_1.assetBundleId
			end
			local _condition = _result_1
			if _condition == nil then
				_condition = BundleGroupNames.ItemUnarmed
			end
			_result:EquipItem(_condition)
		end
	end
	function HeldItem:OnUnEquip()
		self:Log("OnUnEquip")
	end
	function HeldItem:OnCallToActionStart()
		self:Log("OnCallToActionStart")
		if self:HasChargeTime() then
			self:OnChargeStart()
		else
			self:TryUse()
		end
	end
	function HeldItem:OnCallToActionEnd()
		self:Log("OnCallToActionEnd")
		if self.isCharging then
			self:TryChargeUse()
		end
	end
	function HeldItem:OnChargeStart()
		self:Log("OnChargeStart")
		self.isCharging = true
		self.chargeStartTime = TimeUtil:GetServerTime()
	end
	function HeldItem:TryUse()
		self:Log("TryUse IsCooledDown: " .. tostring(self:IsCooledDown()))
		if self:IsCooledDown() then
			self:TriggerUse(0)
			return true
		end
		return false
	end
	function HeldItem:TryChargeUse()
		self:Log("TryChargeUse IsChargedUp: " .. tostring(self:IsChargedUp()))
		if self:IsChargedUp() then
			self:TriggerUse(1)
			return true
		end
		return false
	end
	function HeldItem:TriggerUse(useIndex)
		self:Log("TriggerUse")
		-- Play the use locally
		if RunUtil:IsClient() then
			self:OnUseClient(useIndex)
		elseif RunUtil:IsServer() then
			self:OnUseServer(useIndex)
		end
	end
	function HeldItem:OnUseClient(useIndex)
		self:Log("OnUse Client")
		self.lastUsedTime = TimeUtil:GetServerTime()
		self.isCharging = false
		-- Play the use locally
		local _result = self.entity.anim
		if _result ~= nil then
			_result:PlayItemUse(useIndex)
		end
		local _result_1 = self.meta.itemAssets
		if _result_1 ~= nil then
			_result_1 = _result_1.onUseSoundId
		end
		if _result_1 ~= "" and _result_1 then
			if self.entity:IsLocalCharacter() then
				local _fn = SoundUtil
				local _exp = self.meta.itemAssets.onUseSoundId
				local _object = {}
				local _left = "volumeScale"
				local _condition = self.meta.itemAssets.onUseSoundVolume
				if _condition == nil then
					_condition = 1
				end
				_object[_left] = _condition
				_fn:PlayGlobal(_exp, _object)
			else
				local _fn = SoundUtil
				local _exp = self.meta.itemAssets.onUseSoundId
				local _exp_1 = self.entity.model.transform.position
				local _object = {}
				local _left = "volumeScale"
				local _condition = self.meta.itemAssets.onUseSoundVolume
				if _condition == nil then
					_condition = 1
				end
				_object[_left] = _condition
				_fn:PlayAtPosition(_exp, _exp_1, _object)
			end
		end
	end
	function HeldItem:OnUseServer(useIndex)
		self:Log("OnUse Server")
		-- Update visual state to match client
		self:OnUseClient(useIndex)
	end
	function HeldItem:IsCooledDown()
		local cooldown = self.meta.itemMechanics.cooldownSeconds
		self:Log("Cooldown: " .. tostring(cooldown) .. " Time: " .. tostring(TimeUtil:GetServerTime()) .. " LastUsedTime: " .. tostring(self.lastUsedTime))
		-- no cooldown no startup
		if cooldown <= 0 then
			return true
		end
		-- If the cooldown is down
		return TimeUtil:GetServerTime() > self.lastUsedTime + cooldown - self.serverOffsetMargin
	end
	function HeldItem:IsChargedUp()
		local chargeUpMin = self.meta.itemMechanics.minChargeSeconds
		self:Log("chargeUpMin: " .. tostring(chargeUpMin))
		-- no charge up
		if chargeUpMin <= 0 then
			return true
		end
		-- If we've charged up enough
		return TimeUtil:GetServerTime() - self.chargeStartTime - self.serverOffsetMargin > chargeUpMin
	end
	function HeldItem:HasChargeTime()
		return self.meta.itemMechanics.minChargeSeconds > 0
	end
end
return {
	HeldItem = HeldItem,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Shared/Item/HeldItems/HeldItem.ts
import { AudioManager } from "Shared/Audio/AudioManager";
import { CSArrayUtil } from "Shared/Util/CSArrayUtil";
import { RandomUtil } from "Shared/Util/RandomUtil";
import { Entity } from "../../Entity/Entity";
import {
	Bundle_ItemUnarmed,
	Bundle_ItemUnarmed_SFX,
	BundleGroup,
	BundleGroupNames,
	ReferenceManagerAssets,
} from "../../Util/ReferenceManagerResources";
import { RunUtil } from "../../Util/RunUtil";
import { TimeUtil } from "../../Util/TimeUtil";
import { ItemMeta } from "../ItemMeta";
import { ItemUtil } from "../ItemUtil";

export class HeldItem {
	private serverOffsetMargin = 0.025;
	protected readonly meta: ItemMeta;
	protected readonly entity: Entity;
	protected readonly bundles: BundleGroup | undefined;
	private lastUsedTime = 0;
	private chargeStartTime = 0;
	private isCharging = false;
	protected currentItemGOs: GameObject[] = [];
	protected currentItemAnimations: Animator[] = [];

	constructor(entity: Entity, newMeta: ItemMeta) {
		this.entity = entity;
		this.meta = newMeta;

		//Load the animation references
		if (this.meta.itemAssets?.assetBundleId) {
			this.bundles = ReferenceManagerAssets.bundleGroups.get(this.meta.itemAssets.assetBundleId);
		}
	}

	protected Log(message: string) {
		return;
		let fullMessage = "Entity: " + this.entity.id + " Item: " + this.meta.displayName + " " + message;
		print(fullMessage);
	}

	public OnEquip() {
		this.Log("OnEquip");
		//Load that items animations and play equip animation
		this.entity.anim?.EquipItem(this.meta.itemAssets?.assetBundleId ?? BundleGroupNames.ItemUnarmed);

		//Play the equip sound
		//TODO need to make bundles string accessible for when you dont know the exact bundle you are loading
		let equipPath = this.bundles?.bundles?.get(3)?.filePaths.get(0);
		if (!equipPath) {
			//Load a default equip sound
			equipPath = ReferenceManagerAssets.ItemUnarmed.bundles
				.get(Bundle_ItemUnarmed.SFX)
				?.filePaths.get(Bundle_ItemUnarmed_SFX.Equip);
		}
		if (equipPath) {
			if (this.entity.IsLocalCharacter()) {
				AudioManager.PlayFullPathGlobal(equipPath, {
					volumeScale: 0.5,
				});
			} else {
				AudioManager.PlayFullPathAtPosition(equipPath, this.entity.model.transform.position, {
					volumeScale: 0.2,
				});
			}
		} else {
			error("No default equip sound found");
		}

		//Spawn the accessories graphics
		const accessories = ItemUtil.GetAccessoriesForItemType(this.meta.ItemType);

		this.currentItemAnimations = [];
		this.currentItemGOs = [];
		this.entity.accessoryBuilder.RemoveAccessorySlot(AccessorySlot.LeftHand);
		this.entity.accessoryBuilder.RemoveAccessorySlot(AccessorySlot.RightHand);

		let j = 0;
		for (const accessory of accessories) {
			let added = this.entity.accessoryBuilder.SetAccessory(accessory);
			//Load the animator for the held item if one exists
			for (let go of CSArrayUtil.Convert(added.gameObjects)) {
				this.currentItemGOs[j] = go;
				j++;
				const anim = go.GetComponent<Animator>();
				if (anim) {
					this.currentItemAnimations.push(anim);
				}
			}
		}
	}

	public OnUnEquip() {
		this.Log("OnUnEquip");
		this.currentItemAnimations = [];
		this.currentItemGOs = [];
		this.chargeStartTime = 0;
		this.isCharging = false;
	}

	public OnCallToActionStart() {
		this.Log("OnCallToActionStart");
		if (this.HasChargeTime()) {
			this.OnChargeStart();
		} else {
			this.TryUse();
		}
	}

	public OnCallToActionEnd() {
		this.Log("OnCallToActionEnd");
		if (this.isCharging) {
			this.TryChargeUse();
		}
	}

	protected OnChargeStart() {
		this.Log("OnChargeStart");
		this.isCharging = true;
		this.chargeStartTime = TimeUtil.GetServerTime();
	}

	protected TryUse() {
		this.Log("TryUse IsCooledDown: " + this.IsCooledDown());
		if (this.IsCooledDown()) {
			this.TriggerUse(0);
			return true;
		}
		return false;
	}

	protected TryChargeUse() {
		this.Log("TryChargeUse IsChargedUp: " + this.IsChargedUp());
		if (this.IsChargedUp()) {
			this.TriggerUse(1);
			return true;
		}
		return false;
	}

	protected TriggerUse(useIndex: number) {
		this.Log("TriggerUse");

		//Play the use locally
		if (RunUtil.IsClient()) {
			this.OnUseClient(useIndex);
		} else if (RunUtil.IsServer()) {
			this.OnUseServer(useIndex);
		}
	}

	/** Runs when an item is used. Runs on every client.*/
	protected OnUseClient(useIndex: number) {
		this.Log("OnUse Client");
		this.lastUsedTime = TimeUtil.GetServerTime();
		this.isCharging = false;

		//Play the use locally
		this.entity.anim?.PlayItemUse(useIndex);
		if (this.meta.itemAssets?.onUseSound) {
			if (this.entity.IsLocalCharacter()) {
				AudioManager.PlayGlobal(RandomUtil.FromArray(this.meta.itemAssets.onUseSound), {
					volumeScale: this.meta.itemAssets.onUseSoundVolume ?? 1,
				});
			} else {
				AudioManager.PlayAtPosition(
					RandomUtil.FromArray(this.meta.itemAssets.onUseSound),
					this.entity.model.transform.position,
					{
						volumeScale: this.meta.itemAssets.onUseSoundVolume ?? 1,
					},
				);
			}
		}
	}

	protected PlayItemAnimation(index: number, hold: boolean) {
		for (let i = 0; i < this.currentItemAnimations.size(); i++) {
			let anim = this.currentItemAnimations[i];
			anim.SetBool("Hold", hold);
			anim.Play("Base Layer.Use" + index);
		}
	}

	protected SetItemAnimationHold(hold: boolean) {
		for (let i = 0; i < this.currentItemAnimations.size(); i++) {
			this.currentItemAnimations[i].SetBool("Hold", hold);
		}
	}

	/** Runs when an item is used, server authorized
	 * return true if you can use the item */
	protected OnUseServer(useIndex: number) {
		this.Log("OnUse Server");
		//Update visual state to match client
		this.OnUseClient(useIndex);
	}

	public IsCooledDown(): boolean {
		let cooldown = this.meta.itemMechanics.cooldownSeconds;
		this.Log(
			"Cooldown: " + cooldown + " Time: " + TimeUtil.GetServerTime() + " LastUsedTime: " + this.lastUsedTime,
		);
		//no cooldown no startup
		if (cooldown <= 0) return true;

		//If the cooldown is down
		return TimeUtil.GetServerTime() > this.lastUsedTime + cooldown - this.serverOffsetMargin;
	}

	public IsChargedUp(): boolean {
		let chargeUpMin = this.meta.itemMechanics.minChargeSeconds;
		this.Log("chargeUpMin: " + chargeUpMin);
		//no charge up time
		if (chargeUpMin <= 0) return true;

		//If we've charged up enough
		return TimeUtil.GetServerTime() - this.chargeStartTime - this.serverOffsetMargin > chargeUpMin;
	}

	public HasChargeTime(): boolean {
		return this.meta.itemMechanics.minChargeSeconds > 0;
	}
}
